src/tangent_types/thunks.jl

```diff
@@ -59,12 +59,41 @@ LinearAlgebra.UpperTriangular(a::AbstractThunk) = UpperTriangular(unthunk(a))
 LinearAlgebra.Symmetric(a::AbstractThunk, uplo=:U) = Symmetric(unthunk(a), uplo)
 LinearAlgebra.Hermitian(a::AbstractThunk, uplo=:U) = Hermitian(unthunk(a), uplo)
 
-function LinearAlgebra.diagm(kv::Pair{<:Integer,<:AbstractThunk}...)
-    return diagm((k => unthunk(v) for (k, v) in kv)...)
-end
-function LinearAlgebra.diagm(m, n, kv::Pair{<:Integer,<:AbstractThunk}...)
-    return diagm(m, n, (k => unthunk(v) for (k, v) in kv)...)
+# diagm: not we only dispatch if occurs in first three positions do a few positions,
+# since we will incur type-piracy if we were to  unthunk any mix of thunk and nonthunked 
+# Pairs. In practice using more than 3 is rare. and used can always manually if needed.
+# Basically below we are generating dispatches for each mix of pairs of either
+# AbstractVector or AbstractThunk, for first 3 positions.
+"`unthunk`s the value part of a series of key-value pairs"
+_unthunk_vals(kvs::Pair...) = (k => unthunk(v) for (k, v) in kvs)
```

Comment:
I would have guessed this ought to be an `ntuple` not a generator. But perhaps inlined into a splat it won't matter? Haven't tried to test.